File: met-api/src/met_api/schemas/survey.py
"""Survey schema class.

Manages the survey
"""

from marshmallow import EXCLUDE, Schema, fields
from .engagement import EngagementSchema
from met_api.constants.comment_status import Status
from met_api.schemas.utils import count_comments_by_status


class SurveySchema(Schema):
    """Schema for survey."""

    class Meta:  # pylint: disable=too-few-public-methods
        """Exclude unknown fields in the deserialized output."""

        unknown = EXCLUDE

    id = fields.Int(data_key='id')
    name = fields.Str(data_key='name')
    form_json = fields.Dict(data_key='form_json')
    created_by = fields.Str(data_key='created_by')
    created_date = fields.Str(data_key='created_date')
    updated_by = fields.Str(data_key='updated_by')
    updated_date = fields.Str(data_key='updated_date')
    engagement_id = fields.Str(data_key='engagement_id')
    is_hidden = fields.Bool(data_key='is_hidden')
    is_template = fields.Bool(data_key='is_template')
    engagement = fields.Nested(EngagementSchema)
    comments_meta_data = fields.Method('get_comments_meta_data')
    tenant_id = fields.Str(data_key='tenant_id')

    def get_comments_meta_data(self, obj):
        """Get the meta data of the comments made in the survey."""
        return {
            'total': len(obj.submissions),
            'pending': count_comments_by_status(obj.submissions, Status.Pending.value),
            'approved': count_comments_by_status(obj.submissions, Status.Approved.value),
            'rejected': count_comments_by_status(obj.submissions, Status.Rejected.value),
            'needs_further_review': count_comments_by_status(
                obj.submissions,
                Status.Needs_further_review.value)
        }

File: met-web/src/models/survey.ts
import { FormBuilderData } from 'components/Form/types';
import { createDefaultEngagement, Engagement } from './engagement';

export interface Survey {
    id: number;
    name: string;
    responseCount: number;
    created_date: string;
    is_hidden: boolean;
    engagement?: Engagement;
    form_json?: FormBuilderData;
    comments?: unknown;
    comments_meta_data: SurveyCommentData;
    engagement_id: number;
    engagement_status_id?: number;
}

export interface SurveyCommentData {
    total: number;
    pending: number;
    approved: number;
    rejected: number;
    needs_further_review: number;
}

export interface SurveySubmissionData {
    total: number;
    approved: number;
    rejected: number;
    pending: number;
    needs_further_review: number;
}

export const createDefaultSurvey = (): Survey => {
    return {
        id: 0,
        name: '',
        responseCount: 0,
        created_date: '',
        is_hidden: false,
        engagement: createDefaultEngagement(),
        comments_meta_data: {
            total: 0,
            pending: 0,
            approved: 0,
            rejected: 0,
            needs_further_review: 0,
        },
        engagement_id: 0,
    };
};
